File: linklist_01.py
# 遍历单向链表
import os

class student:
    def __init__(self):
        self.name = ''
        self.Math = 0
        self.English = 0
        self.no = ''
        self.next = None

head = student()
head.next = None
ptr = head
Math_sum = English_sum = num = student_no = 0
select = 0

while select != 2:
    print('(1) add  (2)leave =>')
    try:
        select = int(input('please input your select:'))
    except ValueError:
        print("input error!")
        print('input again:')
    if select == 1:
        new_data = student()
        new_data.name = input('name:')
        new_data.no = input('student id:')
        new_data.Math = eval(input('math grade:'))
        new_data.English = eval(input('English grade:'))
        new_data.next = None
        ptr.next = new_data
        ptr = ptr.next
        num += 1

ptr = head.next
print()
while ptr != None:
    print("name: %s\tstudent ID: %s\tmath grade: %d\tEnglish grade: %d"\
        %(ptr.name, ptr.no, ptr.Math, ptr.English))
    Math_sum += ptr.Math
    English_sum += ptr.English
    student_no += 1
    ptr = ptr.next

if student_no != 0:
    print('==============================================')
    print('average grade: math = %.2f\tenglish = %.2f' % (Math_sum/student_no, English_sum/student_no))

File: c/linklist_01.c
// 遍历单向链表
#include<stdio.h>
#include<malloc.h>
#include<stdlib.h>
#include<string.h>

#define MaxLength 20

typedef struct LinkListNode{
    char name[MaxLength];
    // char *name;
    int student_ID;
    int math;
    int english;
    struct LinkListNode *next;
}llnode;
typedef llnode *llist;

llist init_list(void)
{
    llist list = (llist)malloc(sizeof(llnode));
    if (list == NULL) {
        printf("init error!\n");
        exit(-1);
    }
    else {
        list->next = NULL;
        strcpy(list->name, "list_head");
        list->student_ID = 0;
        list->math = 0;
        list->english = 0;
        return list;
    }
}

llist add_node()
{
    llist new_node = (llist)malloc(sizeof(llnode));
    if (new_node == NULL) {
        printf("add node failed!\n");
        exit(-1);
    }
    else {
        new_node->next = NULL;
        printf("input student name: ");
        scanf("%s", new_node->name);
        printf("input student ID:  ");
        scanf("%d", &new_node->student_ID);
        printf("input math grade:  ");
        scanf("%d", &new_node->math);
        printf("input English grade:  ");
        scanf("%d", &new_node->english);
        return new_node;
    }
}

void travel_list(llist head)
{
    if (head != NULL) {
        llist ptr = head->next;
        float msum = 0, esum = 0, snum = 0;
        while (ptr != NULL) {
            printf("name: %s\tID: %d\tmath: %d\tenglish: %d\n", ptr->name, ptr->student_ID, ptr->math, ptr->english);
            msum += ptr->math;
            esum += ptr->english;
            snum ++;
            ptr = ptr->next;
        }
        if (snum != 0) {
            printf("===================\n");
            printf("average grade: math = %.2f\tenglish = %.2f\n", msum / snum, esum / snum);
        }
    }
}

void free_list(llist head)
{
    llist ptr = head, temp;
    while (ptr != NULL) {
        temp = ptr->next;
        free(ptr);
        ptr = temp;
    }
    head = NULL;
}

int main(void)
{
    llist head = init_list();
    llist ptr = head;
    int select = 0;
    while(select != 2) {
        printf("(1) add node; (2) leave\n");
        scanf("%d", &select);
        if (select == 1) {
            ptr->next = add_node();
            ptr = ptr->next;
        }
    }
    travel_list(head);
    free_list(head);
    return 0;
}
